src/transformers/utils/quantizers.py

```diff
@@ -0,0 +1,879 @@
+from typing import Any, Dict, List, Optional, Tuple, TypeVar, Union
+
+import torch
+from packaging import version
+
+from ..configuration_utils import PretrainedConfig
+from . import (
+    is_accelerate_available,
+    is_auto_awq_available,
+    is_auto_gptq_available,
+    is_bitsandbytes_available,
+    is_optimum_available,
+    logging,
+)
+
+
+if is_bitsandbytes_available():
+    import bitsandbytes as bnb
+
+    from ..pytorch_utils import Conv1D
+
+from .quantization_config import AwqConfig, BitsAndBytesConfig, GPTQConfig, QuantizationConfigMixin, QuantizationMethod
+
+
+logger = logging.get_logger(__name__)
+
+DeviceMap = TypeVar("DeviceMap", str, Dict[str, int])
+QuantizationConfig = TypeVar("QuantizationConfig", QuantizationConfigMixin, Dict)
+StateDict = TypeVar("StateDict", bound=Dict[str, Any])
+MaxMemory = TypeVar("MaxMemory", bound=Dict[Union[int, str], Union[int, str]])
+PreTrainedModel = TypeVar("PreTrainedModel", bound=Any)  # TODO: proper type def
+
+
+class QuantizationStatus(str, Enum):
+    FRESH = "fresh"
+    PREQUANTIZED = "prequantized"
+
+
+class QuantizationConfigParser:
+    """
+    Parser for quantization configuration during model loading. Resolves conflicts between quantization config passed
+    via `from_pretrained` and the one present in the model. Returns a quantizer instance to be used for model loading
+    or None for not-quantized model.
+    """
+
+    def __init__(self):
+        pass
+
+    def parse_config_from_args(self, quantization_config_from_args=None, **kwargs) -> Dict[str, Any]:
+        """
+        Parse quantization config from the arguments passed to `from_pretrained`. Sets the `quantization_config` and
+        `quantization_method` attributes based on args.
+            quantization_config_from_args: The `quantization_config` passed to `from_pretrained`. **kwargs: Additional
+            kwargs passed to `from_pretrained`, like `load_in_8bit`.
+        Returns:
+            Additional kwargs with quantization kwargs popped.
+        """
+        self.quantization_config = quantization_config_from_args
+        # note that we remove `load_in_4bit, load_in_8bit` from kwargs here.
+        load_in_4bit = kwargs.pop("load_in_4bit", False)
+        load_in_8bit = kwargs.pop("load_in_8bit", False)
+        if load_in_4bit or load_in_8bit:
+            warnings.warn(
+                "The `load_in_4bit` and `load_in_8bit` arguments are deprecated and will be removed in the future versions. "
+                "Please, pass a `quantization_config` dict or instance instead.",
+                FutureWarning,
+            )
+
+        self.quantization_method = None
+        if self.quantization_config is not None:
+            # BNB used as default method if Q config present but has no quant_method  # TODO: consider warning / raising error.
+            self.quantization_method = getattr(
+                self.quantization_config, "quant_method", QuantizationMethod.BITS_AND_BYTES
+            )
```

Comment:
Passing a dict as an arg for `quantization_config` is only properly handled for bitsandbytes. Now that we have multiple quantization scheme, we need to make sure that the dict contain a `quantization_method` value.

Reply:
addressed this in `QuantizationConfigParser.validate_quantization_config()`. For now the code issues a warning.

Reply:
I will have a look but in my opinion, we should probably raise an error.

Reply:
changed to raising error. see `QuantizationConfigParser.validate_quantization_config()`

Reply:
reverted to warning in case `load_in_8/4bit` key is True. 
reason: `"ybelkada/bloom-1b7-8bit"` in the 8bit tests has quantization config in dict form.